🚀 2단계 - 문자열 계산기

giyeon95 opened the pull request:
안녕하세요 리뷰어님!

2단계 요구사항 구현 후 리뷰 요청 드립니다.🙏

giyeon95 committed:
docs: 2단계 요구사항 정의

giyeon95 committed:
feat: Operation 기능 구현

giyeon95 committed:
feat: 계산기 전체 기능 구현

ddaaac commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,23 @@
+class Calculator {
+
+    fun calculate(input: String): Int {
+        require(input.isNotBlank()) {
+            throw IllegalArgumentException("빈 값은 올 수 없어요")
+        }
```
코틀린에서 제공하는 [`require(), check()`](https://seosh817.tistory.com/155)을 사용해주셨어요 💯 
다만, 사용법이 살짝 바뀌면 더 좋을 것 같아요. 용례를 확인하고 코드를 수정해볼까요?

ddaaac commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,23 @@
+            throw IllegalArgumentException("빈 값은 올 수 없어요")
+        }
+
+        val values = input.split(" ")
```
의미있는 문자열은 상수로 분리할 수 있을 것 같아요!

giyeon95 replied:
요부분 질문이 있습니다!

Kotlin에서 아래와 같이 상수로 분리하는 방법이 있을 것 같은데요. (전역 상수 제외)

1. 클래스 상단 (파일)에 private const val 로 분리한다.
```
private const val DELIMITERS = " "

class Calculator { ... }
```

2. companion object 내부에 정의 한다.
```
class Calculator {

    companion object {
        private const val DELIMITERS = " "
    }
}
```

2번 방식은 동반객체를 불필요하게 생성하는 것이 아닌가? 라는 의견과, 개인적으로는 2번 보다는 1번의 가독성이 좋아 보여 1번의 방식이 더 낫다고 생각이 드는데요~

리뷰어님은  어떤 의견이신지와 참고할만한 문서가 있을지 궁금합니다.😁

ddaaac replied:
이 부분은 수업시간에 다룰 수도 있을텐데요, 자바 코드로 어떻게 변환되는지 확인해보시면 좀 더 도움이 될 것 같아요.
<img width="348" alt="image" src="https://user-images.githubusercontent.com/43440195/200109215-d1bfa368-3e3d-4a00-96e5-bab4b979cdbf.png">
위와 같은 코틀린 코드는 아래와 같은 자바 코드로 변환됩니다.
<img width="474" alt="image" src="https://user-images.githubusercontent.com/43440195/200109230-ab4f9e07-3d41-4945-87d4-7e6f32fcee03.png">
<img width="400" alt="image" src="https://user-images.githubusercontent.com/43440195/200109237-2d78be46-7102-46f7-b78e-6d1308d17bb6.png">
그래서 Person 클래스에 동반되는/묶이는 성격의 상수라면 동반 객체의 상수로 정의하고, 묶일 필요가 없다면 탑레벨 상수로 정의할 수 있을 것 같아요. [이 스레드](https://stackoverflow.com/a/49977253)도 읽어보면 좋을 것 같네요.

ddaaac commented on src/main/kotlin/calculator/Operation.kt:
```diff
@@ -0,0 +1,18 @@
+
+    companion object {
+        fun of(inputSymbol: String): Operation {
+            return values().find { operation -> operation.symbol == inputSymbol }
+                ?: throw IllegalArgumentException("사칙연산 기호가 아니에요. $inputSymbol")
```
엘비스 연산자 활용 잘 해주셨어요 💯

ddaaac commented on src/main/kotlin/calculator/Operation.kt:
```diff
@@ -0,0 +1,18 @@
+    val symbol: String,
+    val operate: (Int, Int) -> Int
+) {
+    PLUS("+", { a, b -> a.plus(b) }),
```
```suggestion
    PLUS("+", { a, b -> a + b }),
```
연산자 오버로딩 함수를 사용해주셨네요! 위와 같이 나타낼 수도 있어요. 오버로딩된 연산자를 사용할지, 함수를 그대로 사용할지는 그 때 상황에 맞춰 적절히 사용해주시면 될 것 같습니다.

giyeon95 replied:
코틀린의 기능(연산자 오버로딩 함수)를 사용해보자 라는 생각으로 사용해보았습니다만.. 
지금 같은 코드에서는 가독성이 떨어지는 것 같다는 생각도 드네요..😅

ddaaac commented on src/test/kotlin/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,36 @@
+
+class CalculatorTest {
+
+    private lateinit var calculator: Calculator
```
[지연-초기화 프로퍼티](https://kotlinlang.org/docs/properties.html#late-initialized-properties-and-variables)를 잘 사용해주셨네요 💯

ddaaac commented on src/test/kotlin/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,36 @@
+    private lateinit var calculator: Calculator
+
+    @BeforeEach
+    internal fun setUp() {
```
인텔리제이에서 java 테스트 메서드를 만들 때 기본 접근제어자가 package-private이라 여기서는 internal이 사용된 것 같아요.
제 생각엔 그냥 public으로 작성해도 될 거 같아요. [인텔리제이 공식 가이드](https://kotlinlang.org/docs/jvm-test-using-junit.html#add-the-code-to-test-it)에서도 따로 접근제어자를 사용하지 않고 있네요.

giyeon95 replied:
음 제공해주신 공식 가이드 문서를 보니, 클래스 레벨에서 internal을 붙여주고 있기는 하네요.
말씀주신 것처럼 인텔리제이에서 생성 해준 메소드를 그대로 사용 하여서 생겼습니다.😀

다만 internal을 생략하면 조금 더 깔끔해진다는 점에서는 공감을 하지만, internal의 설정을 굳이 다 지워야 할까? 라는 생각도 들기는 합니다.. 오히려 internal이 붙었을때 나쁠 건 없지 않을까 싶기도 하구요.🤔 (intellij에서 생성시 지원하는 설정이 있을것 같기도 하네요.)

ddaaac replied:
저 같은 경우는 자바를 사용하지 않고 코틀린을 사용하고 있기 때문에 템플릿 설정에서 아래처럼 메서드에 public 접근제어자를 붙여줬습니다.
<img width="762" alt="image" src="https://user-images.githubusercontent.com/43440195/200108633-0c0d1fb7-cc27-4d63-a42a-d674562c58d0.png">
필요성을 못 느끼시거나 현업과 미션할 때 번갈아 설정하는 게 불편하시다면 지금과 같이 사용해도 괜찮을 것 같아요 😄

ddaaac commented on src/test/kotlin/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,36 @@
+        assertThat(calculator.calculate(input)).isEqualTo(expected)
+    }
+
+    @EmptySource
+    @ParameterizedTest(name = "`{arguments}` 인경우 예외가 발생한다.")
+    internal fun calculateExceptionTest(input: String) {
+
+        // when, then
+        assertThatIllegalArgumentException().isThrownBy {
+            calculator.calculate(input)
+        }
```
요구사항 중엔 null 또는 빈값에 대한 처리가 명시되어 있습니다. 입력에 null을 추가해볼까요?

ddaaac commented on src/test/kotlin/calculator/OperationTest.kt:
```diff
@@ -0,0 +1,30 @@
+class OperationTest {
+
+    @CsvSource(value = ["+:12", "-:8", "*:20", "/:5"], delimiter = ':')
+    @ParameterizedTest(name = "사칙연산 `{arguments}` 의 결과가 정상적으로 동작한다.")
```
given/when/then 형식의 테스트 이름 너무 좋습니다 💯

ddaaac reviewed (changes requested):
안녕하세요 기연님 👋 

2단계 구현 깔끔하게 잘 해주셨습니다.

소소한 코멘트 몇가지 남겼으니 확인 부탁드려요~

giyeon95 committed:
refactor: require 내 IllegalArgumentException 코드 제거

코틀린에서 제공하는 [`require(), check()`](https://seosh817.tistory.com/155)을 사용해주셨어요 💯
다만, 사용법이 살짝 바뀌면 더 좋을 것 같아요. 용례를 확인하고 코드를 수정해볼까요?

_Originally posted by @ddaaac in https://github.com/next-step/kotlin-racingcar/pull/731#discussion_r1012890831_

giyeon95 committed:
refactor: 의미있는 문자열 상수로 추출

        의미있는 문자열은 상수로 분리할 수 있을 것 같아요!

_Originally posted by @ddaaac in https://github.com/next-step/kotlin-racingcar/pull/731#discussion_r1012892338_

giyeon95 committed:
refactor: 연산자 오버로딩 함수 제거

연산자 오버로딩 함수를 사용해주셨네요! 위와 같이 나타낼 수도 있어요. 오버로딩된 연산자를 사용할지, 함수를 그대로 사용할지는 그 때 상황에 맞춰 적절히 사용해주시면 될 것 같습니다.

_Originally posted by @ddaaac in https://github.com/next-step/kotlin-racingcar/pull/731#discussion_r1012897923_

giyeon95 commented:
꼼꼼한 리뷰 감사합니다.🙏

리뷰 주신 내용 수정 후 재 리뷰 요청 드립니다.

남겨주신 리뷰중에 궁금한 의견을 몇가지 남겨두었습니다.
한번 확인 부탁 드려요.🙇‍♂️

ddaaac reviewed (approved):
안녕하세요 기연님 👋 

리뷰 반영 잘 해주셨습니다. 질문해주신 부분에 대해서는 코멘트 달았어요.

다음 단계 진행해주세요!
